### PR TITLE
Refactor and fixes parts of the websocket system

### PR DESCRIPTION
**Pull request by GabrielFleischer**
Refactor the websocket system to be more general and comprehensible

Fix a problem when closing the sockets

### CODE REVIEW

File: app/src/main/java/com/github/dedis/student20_pop/utility/network/IMessageListener.java [hunk 1]
```diff
@@ -0,0 +1,14 @@
+package com.github.dedis.student20_pop.utility.network;
+
+/**
+ * Listener of received messages by the websocket
+ */
+public interface IMessageListener {
```
Comment (jbmichel14) [inline app/src/main/java/com/github/dedis/student20_pop/utility/network/IMessageListener.java:6]: Just by curiosity, why is it called IMessageListnener (why the I) ?

**Review by jbmichel14 — APPROVED** [review]

File: app/src/main/java/com/github/dedis/student20_pop/utility/network/WebSocketLowLevelProxy.java [hunk 2]
```diff
@@ -240,29 +214,32 @@ public void purge() {
...
                 try {
                     s.close();
                 } catch (IOException e) {
                     e.printStackTrace();
                 }
             });
         }
+
+        // Complete all pending requests and remove them
+        Iterator<Map.Entry<Integer, RequestEntry>> it = requests.entrySet().iterator();
```
Comment (celinecamacho) [inline app/src/main/java/com/github/dedis/student20_pop/utility/network/WebSocketLowLevelProxy.java:238]: What does the iterator `it` represent? Maybe you can rename it to something more helpful

File: app/src/main/java/com/github/dedis/student20_pop/utility/protocol/LowLevelProxy.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+     *
+     * @param channel to subscribe to
+     * @return a completable future holding the response value
+     */
+    CompletableFuture<Integer> unsubscribe(String channel);
+
+    /**
+     * Publish an event on given channel
+     *
+     * @param channel to publish the event on
```
Comment (celinecamacho) [inline app/src/main/java/com/github/dedis/student20_pop/utility/protocol/LowLevelProxy.java:35]: The comments for the parameters `sender` and `key` are missing

**Review by celinecamacho — CHANGES_REQUESTED** [review]
Great, just one comment missing. Also I think it would be helpful to rename a variable

File: app/src/main/java/com/github/dedis/student20_pop/utility/network/WebSocketLowLevelProxy.java (hunk 2 above)
Comment (GabrielFleischer) [reply app/src/main/java/com/github/dedis/student20_pop/utility/network/WebSocketLowLevelProxy.java:238]: it for iterator

File: app/src/main/java/com/github/dedis/student20_pop/utility/network/IMessageListener.java (hunk 1 above)
Comment (GabrielFleischer) [reply app/src/main/java/com/github/dedis/student20_pop/utility/network/IMessageListener.java:6]: For Interface, in some convention, we always add an 'I' before an interface name. I am too used to this convention.
I'm changing it for concistency

File: app/src/main/java/com/github/dedis/student20_pop/utility/protocol/LowLevelProxy.java (hunk 3 above)
Comment (GabrielFleischer) [reply app/src/main/java/com/github/dedis/student20_pop/utility/protocol/LowLevelProxy.java:35]: Added in latest commit

**Review by celinecamacho — APPROVED** [review]

**Pull request merged**